src/sources/logplex.rs

```diff
@@ -0,0 +1,190 @@
+            });
+
+        let ping = warp::get2().and(warp::path("ping")).map(|| "pong");
+
+        let routes = svc.or(ping);
+
+        let (_, server) = warp::serve(routes).bind_with_graceful_shutdown(self.address, tripwire);
+
+        Ok(Box::new(server))
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn source_type(&self) -> &'static str {
+        "logplex"
+    }
+}
+
+fn body_to_events(body: impl Buf) -> Vec<Event> {
+    let rdr = BufReader::new(body.reader());
+    rdr.lines()
+        .filter_map(|res| {
+            res.map_err(|error| error!(message = "Error reading request body", ?error))
+                .ok()
+        })
+        .filter(|s| !s.is_empty())
+        .map(line_to_event)
+        .collect()
+}
+
+fn line_to_event(line: String) -> Event {
+    let parts = line.splitn(8, ' ').collect::<Vec<&str>>();
+
+    if parts.len() == 8 {
+        let timestamp = parts[2];
+        let hostname = parts[3];
+        let app_name = parts[4];
+        let proc_id = parts[5];
+        let message = parts[7];
+
+        let mut event = Event::from(message);
+        let log = event.as_mut_log();
+
+        if let Ok(ts) = timestamp.parse::<DateTime<Utc>>() {
+            log.insert(event::TIMESTAMP.clone(), ts);
+        }
+
+        log.insert(event::HOST.clone(), hostname);
+
+        log.insert("app_name", app_name);
+        log.insert("proc_id", proc_id);
+
+        event
+    } else {
+        warn!(
+            message = "Line didn't match expected logplex format. Forwarding raw message.",
+            fields = parts.len()
+        );
+        Event::from(line)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::LogplexConfig;
+    use crate::{
+        event::{self, Event},
+        runtime::Runtime,
+        test_util::{self, collect_n},
+        topology::config::{GlobalOptions, SourceConfig},
+    };
+    use chrono::{DateTime, Utc};
+    use futures::sync::mpsc;
+    use http::Method;
+    use pretty_assertions::assert_eq;
+    use std::net::SocketAddr;
+
+    fn source(rt: &mut Runtime) -> (mpsc::Receiver<Event>, SocketAddr) {
+        test_util::trace_init();
+        let (sender, recv) = mpsc::channel(100);
+        let address = test_util::next_addr();
+        rt.spawn(
+            LogplexConfig { address }
+                .build("default", &GlobalOptions::default(), sender)
+                .unwrap(),
+        );
+        (recv, address)
+    }
+
+    fn send(address: SocketAddr, body: &str) -> u16 {
+        let len = body.lines().count();
+        reqwest::Client::new()
+            .request(Method::POST, &format!("http://{}/events", address))
+            .header("Logplex-Msg-Count", len)
+            .header("Logplex-Frame-Id", "frame-foo")
+            .header("Logplex-Drain-Token", "drain-bar")
+            .body(body.to_owned())
+            .send()
+            .unwrap()
+            .status()
+            .as_u16()
+    }
+
+    #[test]
+    fn router_log() {
```

Comment:
Would it make sense to test `line_to_event` a bit more? I wonder if we could fuzz or quickcheck at the most extreme too (not now but in the future).

Reply:
Yeah, it would. I've deliberately kept it pretty simple for now and provided fallbacks for when assumptions aren't met, but it would definitely be a good thing to fuzz.

In the meantime, I'll push a couple of other simple unit tests for it.